### DISCUSSION TITLE
Using store ouside components discard a config variable

### DISCUSSION THREAD

**Discussion by 0xmemorygrinder** [pmndrs/zustand#2106] [Bug report]
Hello, I am using zustand for a frontend project that needs to be generic and only rely on its config variables defined in the project.

Here is a sample config variable I use : 
```typescript
// src/config/vaults.config.ts:

export const vaultsConfig: VaultConfig = {
    key1: {
        subKey1: value1,
        ...
    },
    key2: ...
}
```

When I create a slice of my store, I use this config to create my state data. 
```typescript
function createDefaultData(): [VaultId, Vault][] {
  console.log(vaultsConfig);
  return Object.entries(vaultsConfig).map(([vaultId, vaultConfig]) => {
    return // prepared data
  })
}

const defaultState: State = {
  vaults: new Map<VaultId, Vault>(createDefaultData())
}

export const createVaultsStore: StateCreator<Store, [], [], VaultsStore> = (set, get) => ({
  ...defaultState,
  //methods here
});
```
I then create my store with all the slices that compose it.

Everything works fine when using the store inside next components. However, I need to access and mutate the state ouside of react components (which is normally supported.

I get an error when accessing the state by using 
```typescript
useStore().getState().myMethodToCall()
```

At this point, the error tells me that the config passed to Object.entries during my store creation is undefined. I logged the value and it really is.

How can I fix this ? 

Thanks for your time.

**Comment by dbritto-dev** [comment]
@0xMemoryGrinder would you mind sharing us a codesandbox demo? Just a minimal repto I would like to help you

**Reply by 0xmemorygrinder** [reply]
I created a minimal example of my architecture in codesandbox but its workls like a charm. On my end, I get an error as soon as I call useStore, even without argument behind.

**Reply by 0xmemorygrinder** [reply]
What I do not understand to be exact is why a variable (vaultsConfig) that has nothing to do with zustand become undefined as soon as I use my store outside react @dbritto-dev

**Reply by dbritto-dev** [reply]
@0xMemoryGrinder are you using `create` or `createStore`?

**Reply by 0xmemorygrinder** [reply]
@dbritto-dev create why ?

**Reply by dbritto-dev** [reply]
@0xMemoryGrinder how can I replicate your issue? without that I can't help you

**Reply by 0xmemorygrinder** [reply]
@dbritto-dev This is precisely my problem, I tried to replicate the issue on codesandbox and it works normally. I do not find the difference between my example and my project. 
Here is the sample architecture I got : https://codesandbox.io/p/sandbox/distracted-glade-w72hpm
Any idea on where I should look in priority in my project ?

**Reply by dbritto-dev** [reply]
@0xMemoryGrinder now I can see couple things that you didn't mention in this thread:
- You are using nextjs
- You are using app router 

If you are calling any method outside of react on server components you can expect that behavior. Due to how you are chaining the deps: config -> service -> store -> config so what you are facing is a circular dependency that means that sometimes you are accessing the store without having the config ready

> **Note:** you need to be careful while using server and client components

**Reply by 0xmemorygrinder** [reply]
@dbritto-dev Thanks for the hints, however all my components are rendered client side (since nearly all of them use a client side only library).
I understand tour concern about circular dependency but in this case, can you explain me why its works on the codesandbox ? 

Thanks for your answer anyway

**Reply by dbritto-dev** [reply]
@0xMemoryGrinder actually, codesandbox if works on codesandbox doesn't mean that should works on a real environment but an easy test could be importing `vaultsConfig` on your layout and see what happen, however we can ensure that everything works on app router I would try to use `createStore` instead of `create` since that is completely separated from react and you can connect the store later with `useStore` or `use-zustand` (the preferred way to use zustand in nextjs)

**Reply by 0xmemorygrinder** [reply]
Ok I will try it asap. Do you have any link concerning the usage of `createStore` ? 

Thank you very much for all your answers

**Reply by dbritto-dev** [reply]
@0xMemoryGrinder sure, here you go:
- https://github.com/dai-shi/use-zustand
- https://docs.pmnd.rs/zustand/guides/initialize-state-with-props

**Reply by 0xmemorygrinder** [reply]
Hello @dbritto-dev, 

I fixed the two errors I got (the circular dependency and using createStore). I now got a weird behavior : My state cannot be used in any builtin react hook (useEffect, useRef...).
The component is now rerendered with the new value but if I use a useEffect on the stored value, the callback is never called. I even tried to implement a custom useEffect logic with useRef (and even useState) but my comparison always returns true as the value gets updated before I compare it.

Here is a hook I created to replicate the useEffect : 
```typescript
import VaultMode from "@/types/vault/VaultMode";
import VaultId from "@/types/vault/VaultId";
import {useEffect, useRef, useState} from "react";
import {appStore} from "@/store";
import {deepEqual} from "@wagmi/core";
import useVaultService from "@/hooks/vault/useVaultService";
import {useZustand} from "use-zustand";

function isDeepEqual(a: any, b: any) {
  const aKeys = Object.keys(a);
  const bKeys = Object.keys(b);
  console.log('a', a, 'b', b);
  if (aKeys.length != bKeys.length) return false;
  for (let i = 0; i < aKeys.length; i++) {
    const key = aKeys[i];
    if (a[key] != b[key]) return false;
  }
  return true;
}

export default function useInputAmountsEffect(vault: VaultId, selectionIdx: number, mode: VaultMode) {
  const inputAmounts = useZustand(appStore, state => mode == "deposit" ? state.getVaultDepositInputAmounts(vault, selectionIdx) : state.getVaultWithdrawInputAmounts(vault, selectionIdx)); // inputAmounts is updated in a component
  const service = useVaultService(vault);
  const [previous, setPrevious] = useState<{[p: string]: bigint} | null>(null);
  console.log('previous', previous); // This keeps logging the new value even if the if below is never taken

  if (previous == null || !isDeepEqual(inputAmounts, previous)) {
    console.log('changing');
    setPrevious(inputAmounts);
    service.computeDepositOutputTokenAmounts(inputAmounts) 
  }
}
```

Do you possibly know why I get this ? 

Thanks for your answer

**Reply by dbritto-dev** [reply]
@0xMemoryGrinder sure I'll look into that

**Reply by 0xmemorygrinder** [reply]
Hello, do you have any update ?

**Reply by dbritto-dev** [reply]
@0xMemoryGrinder can you try to recreate a minimal repro? I wasn't able to replicate your issue

**Reply by 0xmemorygrinder** [reply]
I will try later today. If I cannot reproduce I will invite you to my repository

**Reply by 0xmemorygrinder** [reply]
After days and days of debugging, I found a solution for my problem ! 

If you store an object in zustand and retrieve the plain object, you need to make a plain copy of the object before using it in a ref of useState otherwise it points to the same address in memory and the value stored in ref or state is updated when you update your store.

**Reply by dbritto-dev** [reply]
> After days and days of debugging, I found a solution for my problem ! 
> 
> 
> 
> If you store an object in zustand and retrieve the plain object, you need to make a plain copy of the object before using it in a ref of useState otherwise it points to the same address in memory and the value stored in ref or state is updated when you update your store. 
> 
> 

@0xMemoryGrinder yeah, that's the expected behavior
